### ISSUE TITLE
tegra-bootloader-update segfault

### ISSUE THREAD

**Issue by rbuchmann77** [OE4T/tegra-demo-distro#304]
Hello,

Since I updated my project based on tegra-demo-distro from JP **4.6.2** to JP 4.6.5 (l4t r32.7.5), mender install is broken and I would appreciate any help/support for this.

"mender install" failed on my board. I can reproduce the error as follow:

```
# /var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update
Installing NVIDIA bootloader update payload
Current boot slot: 0
Native TNSPEC:   3636-300-0001-H.0-1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1
Compatible with: 3636--0001--1--jetson-xavier-nx-devkit-tx2-nx-
Warning: ordered entry list mismatch
  Processing bpmp-fw_b... [OK]
  Processing sce-fw_b... [OK]
  Processing cpu-bootloader_b... [OK]
  Processing spe-fw_b... [no update needed]
  Processing adsp-fw_b... [OK]
  Processing secure-os_b... [OK]
  Processing BMP_b... [OK]
  Processing mts-bootpack_b... [OK]
  Processing mts-preboot_b... [no update needed]
  Processing sc7_b... [OK]
  Processing bpmp-fw-dtb_b... [OK]
  Processing bpmp-fw-dtb_b... [OK]
  Processing bootloader-dtb_b... [OK]
  Processing bootloader-dtb_b... [OK]
  Processing VER_b... [OK]
  Processing VER_b... [OK]
  Processing MB1_BCT_b... [no update needed]
  Processing MB1_BCT_b... [no update needed]
  Processing kernel_b... [OK]
  Processing kernel-dtb_b... [OK]
  Processing kernel-dtb_b... [OK]
  Processing mb2_b... [no update needed]
  Processing BCT... [offset=3584,no update needed]...[offset=16384,no update needed]...[offset=0,no update needed]...[OK]
  Processing mb1_b... [no update needed]
  Processing (null)... /var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update: line 58: 37192 Segmentation fault      chroot "${mnt}" /usr/bin/tegra-bootloader-update /opt/ota_package/bl_update_payload
ERR: bootloader update failed
```
I noticed the warning: "ordered entry list mismatch".
I guessed it was related to a change to partitionning. So I tried to compare it. To do that, I uncompress *tegraflash.tar.gz and run using:

`./BUP_generator.py -c ../build_carrier-adv-tx2-nx-dev/tmp/deploy/images/jetson-xavier-nx-devkit-tx2-nx/Image-initramfs-jetson-xavier-nx-devkit-tx2-nx.bup-payload`

Before merging :
```
BLOB PATH:
Image.gz-initramfs-jetson-xavier-nx-devkit-tx2-nx.bl_only.bup-payload

BLOB HEADER:
       Magic: NVIDIA__BLOB__V2
     Version: v2.1-2021.6-0 (0x01020621)
   Blob Size: 5,035,455 bytes
 Header Size: 48 bytes
 Entry Count: 22 partition(s)
        Type: 0 (0 for update, 1 for BMP)
Uncompressed
   Blob Size: 5,035,455 bytes
   Accessory:
0x0000000000000000

ENTRY TABLE:
|  part_name    | offset |part_size |version |op_mode |                           tnspec                           |
|       bpmp-fw |   2688 |   534816 |  3272  |   0    |                                                            |
|        sce-fw | 537504 |   126032 |  3272  |   0    |                                                            |
|cpu-bootloader | 663536 |   320464 |  3272  |   0    |                                                            |
|           mb2 | 984000 |   122784 |  3272  |   0    |                                                            |
|        spe-fw |1106784 |    95984 |  3272  |   0    |                                                            |
|       adsp-fw |1202768 |    77616 |  3272  |   0    |                                                            |
|     secure-os |1280384 |   403264 |  3272  |   0    |                                                            |
|           BMP |1683648 |   207188 |  3272  |   0    |                                                            |
|           mb1 |1890836 |    99904 |  3272  |   2    |                                                            |
|  mts-bootpack |1990740 |  2236640 |  3272  |   2    |                                                            |
|   mts-preboot |4227380 |    58384 |  3272  |   2    |                                                            |
|           sc7 |4285764 |    23632 |  3272  |   2    |                                                            |
|   bpmp-fw-dtb |4309396 |   113600 |  3272  |   0    |3636-300-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|   bpmp-fw-dtb |4422996 |   113600 |  3272  |   0    |3636-301-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|bootloader-dtb |4536596 |   196112 |  3272  |   0    |3636-300-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|bootloader-dtb |4732708 |   196112 |  3272  |   0    |3636-301-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|           VER |4928820 |      102 |  3272  |   0    |3636-300-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|           VER |4928922 |      101 |  3272  |   0    |3636-301-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|           BCT |4929023 |     3584 |  3272  |   2    |3636-300-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|           BCT |4932607 |     3584 |  3272  |   2    |3636-301-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|       MB1_BCT |4936191 |    49632 |  3272  |   0    |3636-300-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|       MB1_BCT |4985823 |    49632 |  3272  |   0    |3636-301-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
```

After merging: 
```
BLOB PATH:
Image-initramfs-jetson-xavier-nx-devkit-tx2-nx.bup-payload

BLOB HEADER:
       Magic: NVIDIA__BLOB__V2
     Version: v2.1-2021.6-0 (0x01020621)
   Blob Size: 31,069,720 bytes
 Header Size: 48 bytes
 Entry Count: 25 partition(s)
        Type: 0 (0 for update, 1 for BMP)
Uncompressed
   Blob Size: 31,069,720 bytes
   Accessory:
0x0000000000000000

ENTRY TABLE:
|  part_name    | offset  |part_size |version |op_mode |                           tnspec                           |
|       bpmp-fw |    3048 |   534816 |  3275  |   0    |                                                            |
|        sce-fw |  537864 |   126032 |  3275  |   0    |                                                            |
|cpu-bootloader |  663896 |   312528 |  3275  |   0    |                                                            |
|           mb2 |  976424 |   122784 |  3275  |   0    |                                                            |
|        spe-fw | 1099208 |    95984 |  3275  |   0    |                                                            |
|       adsp-fw | 1195192 |    77616 |  3275  |   0    |                                                            |
|     secure-os | 1272808 |   407360 |  3275  |   0    |                                                            |
|           BMP | 1680168 |   207188 |  3275  |   0    |                                                            |
|           mb1 | 1887356 |    99904 |  3275  |   2    |                                                            |
|  mts-bootpack | 1987260 |  2197920 |  3275  |   2    |                                                            |
|   mts-preboot | 4185180 |    58256 |  3275  |   2    |                                                            |
|           sc7 | 4243436 |    23632 |  3275  |   2    |                                                            |
|   bpmp-fw-dtb | 4267068 |   113600 |  3275  |   0    |3636-300-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|   bpmp-fw-dtb | 4380668 |   113600 |  3275  |   0    |3636-301-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|bootloader-dtb | 4494268 |   196112 |  3275  |   0    |3636-300-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|bootloader-dtb | 4690380 |   196112 |  3275  |   0    |3636-301-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|           VER | 4886492 |      102 |  3275  |   0    |3636-300-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|           VER | 4886594 |      102 |  3275  |   0    |3636-301-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|           BCT | 4886696 |     3584 |  3275  |   2    |3636-300-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|           BCT | 4890280 |     3584 |  3275  |   2    |3636-301-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|       MB1_BCT | 4893864 |    49632 |  3275  |   0    |3636-300-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|       MB1_BCT | 4943496 |    49632 |  3275  |   0    |3636-301-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|        kernel | 4993128 | 25684368 |  3275  |   0    |                                                            |
|    kernel-dtb |30677496 |   196112 |  3275  |   0    |3636-300-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
|    kernel-dtb |30873608 |   196112 |  3275  |   0    |3636-301-0001--1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1 |
```
The order seems identical except: 
* It added 1x kernel, 2x kernel-dtb.
* offset/part_size are different

I compared bitbake env before/after merging (using bitbake -e linux-tegra and linux-tegra-initrd-flash, because the recipe was splitted since JP465). I failed to find a misconfiguration or any useful idea.

Is it possible to have a compatible mender file to upgrade from a JP4.6.4 ?
How to fix ArtifactInstall_Leave_80_bl-update/partitionning?

**Comment by madisongh** [comment]
The problem might be the duplicate entries in the BUP.  The update tool is reporting
```
Native TNSPEC:   3636-300-0001-H.0-1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1
Compatible with: 3636--0001--1--jetson-xavier-nx-devkit-tx2-nx-
```
so there shouldn't be entries in the BUP package for both FAB `300` and `301`. Did you customize the `TEGRA_BUPGEN_SPECS` variable somewhere to add an entry for `fab=301` for that machine?  The machine config we have in the layer doesn't have that, just a single entry for `fab=300`.

**Comment by rbuchmann77** [comment]
You are right. We had the following append:
`TEGRA_BUPGEN_SPECS:append:jetson-xavier-nx-devkit-tx2-nx = " fab=301;boardsku=0001;boardrev="`

With this append,  "bitbake -e" gives:
`TEGRA_BUPGEN_SPECS="fab=300;boardsku=0001 fab=301;boardsku=0001;boardrev="`

I think this is not a duplicate entry in TEGRA_BUPGEN_SPECS.

If I remove this append, mender install now exit successfully (= exit 0). Unfortunately, after reboot, the system boots on the previous/wrong slot (0) : 

```
...
0%                              50%                              100%
|---------------------------------|---------------------------------|
time="2024-07-30T08:23:54Z" level=info msg="Installer: authenticated digital signature of artifact"
time="2024-07-30T08:23:54Z" level=info msg="Opening device \"/dev/mmcblk0p33\" for writing"
time="2024-07-30T08:23:54Z" level=info msg="Native sector size of block device /dev/mmcblk0p33 is 512 bytes. Mender will write in chunks of 1048576 bytes"
...
time="2024-07-30T08:24:58Z" level=info msg="Enabling partition with new image installed to be a boot candidate: 33"
time="2024-07-30T08:24:58Z" level=info msg="Output (stderr) from command \"fw_printenv\": ERR: no such variable: mender_check_saveenv_canary"
time="2024-07-30T08:24:58Z" level=info msg="Output (stderr) from command \"fw_printenv\": ERR: no such variable: mender_check_saveenv_canary"
time="2024-07-30T08:24:58Z" level=info msg="Executing script: ArtifactInstall_Leave_80_bl-update"
time="2024-07-30T08:24:58Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\": Installing NVIDIA bootloader update payload"
time="2024-07-30T08:24:58Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\": Current boot slot: 0"
time="2024-07-30T08:24:58Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\": Native TNSPEC:   3636-300-0001-H.0-1-0-jetson-xavier-nx-devkit-tx2-nx-mmcblk0p1"
time="2024-07-30T08:24:58Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\": Compatible with: 3636--0001--1--jetson-xavier-nx-devkit-tx2-nx-"
time="2024-07-30T08:24:58Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing bpmp-fw_b... [OK]"
time="2024-07-30T08:24:58Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing sce-fw_b... [OK]"
time="2024-07-30T08:24:58Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing cpu-bootloader_b... [OK]"
time="2024-07-30T08:24:58Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing spe-fw_b... [no update needed]"
time="2024-07-30T08:24:58Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing adsp-fw_b... [OK]"
time="2024-07-30T08:25:00Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing secure-os_b... [OK]"
time="2024-07-30T08:25:00Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing BMP_b... [OK]"
time="2024-07-30T08:25:00Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing mts-bootpack_b... [OK]"
time="2024-07-30T08:25:00Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing mts-preboot_b... [no update needed]"
time="2024-07-30T08:25:00Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing sc7_b... [OK]"
time="2024-07-30T08:25:00Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing bpmp-fw-dtb_b... [OK]"
time="2024-07-30T08:25:01Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing bootloader-dtb_b... [OK]"
time="2024-07-30T08:25:01Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing VER_b... [OK]"
time="2024-07-30T08:25:02Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing MB1_BCT_b... [no update needed]"
time="2024-07-30T08:25:02Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing kernel_b... [OK]"
time="2024-07-30T08:25:02Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing kernel-dtb_b... [OK]"
time="2024-07-30T08:25:02Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing mb2_b... [no update needed]"
time="2024-07-30T08:25:02Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing BCT... [offset=3584,no update needed]...[offset=16384,no update needed]...[offset=0,no update needed]...[OK]"
time="2024-07-30T08:25:02Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing mb1_b... [no update needed]"
time="2024-07-30T08:25:02Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\":   Processing mb1... [no update needed]"
time="2024-07-30T08:25:02Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\": Slot 1 marked as active for next boot"
time="2024-07-30T08:25:02Z" level=info msg="Output (stdout) from command \"/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update\": Successful bootloader update\n"
Use 'commit' to update, or 'rollback' to roll back the update.
At least one payload requested a reboot of the device it updated.

# reboot

# tegra-boot-control -s
Redundancy: enabled (full)
Current slot: 0
Slot 0: priority: 15, suffix: _a, retry_count: 7, successful: YES <= still boot on slot 0
Slot 1: priority: 14, suffix: _b, retry_count: 0, successful: NO
```

So we have now 2 problems:

- We need to support 2 modules "fab 300" and "fab 301". Like in the following post: https://github.com/OE4T/meta-tegra/issues/1067#issue-1363302444 . 
- mender install exit 0 but the system does not boot the new SW.

**Comment by rbuchmann77** [comment]
I managed to save logs of the first boot after updating (fab=300 only):
[Uploading unexpected_slot.log…]()

The first lines do not show up when booting directly to a "successful" slot. I guess something goes wrong in cboot: 
```
ERROR: Highest Layer Module = 0x40, Lowest Layer Module = 0x40,
Aux Info = 0x0, Reason = 0xd
```

I also compared BUP_generator.py's outputs:
* JP 4.6.2 with fab=300 & fab=301 : has some duplicated entries (1 MB1_BCT for 300 and 1 for 301). It was working fine.
* JP 4.6.5 with fab=300 & fab=301 : has also duplicated entries but "mender install" failed.
* JP 4.6.5 with fab=300 only : has no duplicated entries. "mender install" succeeded. However the next slot is NOT successful.

**Comment by madisongh** [comment]
We addressed [the module compatibility issue](https://github.com/OE4T/meta-tegra/issues/1067) you mention by implementing the compatibility specs described there (see e75ee2987208c0ad6f7c7e46a3d12a4b88d71c43). Your 4.6.2/32.7.2-based builds may not have had that change.

> I managed to save logs of the first boot after updating (TEGRA_BUPGEN_SPECS="fab=300;boardsku=0001" on a fab=300 board/soc):

Unfortunately, the log didn't actually get attached. The module ID in that error message doesn't show up in the cboot sources, so it's not clear whether it's cboot or one of the other early bootloaders (mb1/mb2) that's getting the error. Could you try uploading it again?

I've compared what's in the `kirkstone-l4t-r32.7.x` branch against the stock L4T R32.7.5 kit, and I found one discrepancy in our TEGRA_BUPGEN_SPECS setting, which I've just pushed a fix for. That's a BOARDSKU difference, rather than a FAB difference, though, so I'm not sure it would necessarily help here.

It might also help to try USB flashing one of the boards that you're seeing this problem on, to see if flashing from scratch results in a working boot. If it does, then comparing what was flashed onto the board vs. what was updated via the BUP would tell us something.

**Comment by rbuchmann77** [comment]
Here the missing file: [unexpected_slot.log](https://github.com/user-attachments/files/16482336/unexpected_slot.log)

**Comment by madisongh** [comment]
Looks like it's the mb1 loader that's failing there, so it's likely to be a problem with the contents of the MB1, MB1_BCT, or BCT entries in the update payload. That seems strange, though, because from the mender log you posted, it looks like none of those actually got updated (from the 'no update needed' message in that log for those partitions). That's puzzling.

If you can successfully USB-flash a 32.7.5-based build onto your target, and it boots OK, then doing a side-by-side comparison between the flashing process and the BUP build - looking for any differences in which binaries or `.cfg` files got used for each - could help with tracking down the problem.  If even flashing from scratch results in a boot failure, then try comparing what got flashed  with 32.7.5 against your last-known-good flashing (32.7.2, I guess) to see what the differences are.

**Comment by rbuchmann77** [comment]
Thank you for your reply.

I successfully flashed my distro based on JP4.6.5 (on an old board/SOM). Mender install works "fine" on top of it.

Note new SOM (https://forums.developer.nvidia.com/t/pcn210541-micron-dram/291776) still does not work with **my** JP4.6.5. I will try to work on it as soon as I get logs.

I tried several flashing & mender install combo:
* Flash JP4.6.2, then mender install **JP4.6.5** => NOK (1st message of this topic)
* Flash JP4.6.5, then mender install JP4.6.5 => OK
* Flash JP4.6.5, then mender install **JP4.6.2** => OK. Note mender install **JP4.6.5** still does not work.

Mender install **JP4.6.5** always succeed, logs the following and go back to the previous slot:
```
ERROR: Highest Layer Module = 0x40, Lowest Layer Module = 0x40,
Aux Info = 0x0, Reason = 0xd
```

I'm not sure about what you mean by "side-by-side comparison". I did diffs from 2 tegraflash.tar.gz:
All cfg are the same except:
* jetson-xavier-nx-devkit-tx2-nx.cfg (only in JP4.6.2)
* tegra186-mb1-bct-memcfg-p3636-0001-a01.cfg  tegra186-mb1-bct-memcfg-p3636-0003-a01.cfg (only in JP4.6.5)

[cfg.tar.gz](https://github.com/user-attachments/files/16694431/cfg.tar.gz)

I guess we expected changes in BCT since PCN210541.

I do not know how MB1 loader works and how it's configured. I have no idea how to go further.

So, to summarize:
* I need to boot new SOM (FAB302 ?). We are working to get logs. Do I need to append something like TEGRA_BUPGEN_SPECS:append:jetson-xavier-nx-devkit-tx2-nx = " fab=302" ?
* I would like to install JP4.6.5 onto a JP4.6.2.

**Comment by rbuchmann77** [comment]
TEGRA_MENDER_RESERVED_SPACE_MB is implied into the following error:

```
ERROR: Highest Layer Module = 0x40, Lowest Layer Module = 0x40,
Aux Info = 0x0, Reason = 0xd
```

TEGRA_MENDER_RESERVED_SPACE_MB changed from 1024 to 1974 in https://github.com/OE4T/tegra-demo-distro/commit/8d8019ca0db7d32a6a8d1b8efa5d675f3c4e79d3

I reverted to 1024 and it works fine.

I'm not sure to understand all the consequences. Do you think it's safe?

**Comment by madisongh** [comment]
> I reverted to 1024 and it works fine. Unlike to commit's comment, it breaks the compatibility. In my case, it does exactly the opposite.

If your targets were originally flashed with a storage layout that used that 1024 setting, then keeping it at that value sounds like the right solution.

**Comment by rbuchmann77** [comment]
Thank you for the advices.
